### PR TITLE
Refactoring BadBlock

### PR DESCRIPTION
**Pull request by hqjang-pepper**
## Proposed changes

- fix BadBlock log print from multi-line into single line.
- Change BadBlock storing location from Cache to DB.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -2362,57 +2357,42 @@ type BadBlockArgs struct {
...
-Chain config: %v
-
-Number: %v
-Hash: 0x%x
-%v
-
-Error: %v
-##############################
-`, bc.chainConfig, block.Number(), block.Hash(), receiptString, err))
+	logger.Error(fmt.Sprintf(`########## BAD BLOCK ######### Chain config: %v Number: %v Hash: 0x%x %v Error: %v`, bc.chainConfig, block.Number(), block.Hash(), receiptString, err))
```
Comment (ian0371) [inline blockchain/blockchain.go:2395]: Shouldn't this be something like below for better parsing?
```
logger.Error("########## BAD BLOCK #########", "chainConfig", bc.chainConfig, "number", block.Number(), "hash", block.Hash(), "receipt", receiptString, "err", err))
```

Comment (hqjang-pepper) [reply blockchain/blockchain.go:2395]: This should be good though. But `receiptString` is often returned as an empty space. 
So it might seem strange if it's empty. How do you think?

**Comment by hqjang-pepper** [comment]
Adding testcode. This code is draft.

**Comment by hqjang-pepper** [comment]
Test code is updated.

File: blockchain/blockchain.go (hunk 1 above)
Comment (ian0371) [reply blockchain/blockchain.go:2395]: I think it's ok because empty strings are distinguishable (i.e., `receipt= err=...`)
My initial thought was this was for the sake of string parsability.

**Comment by hqjang-pepper** [comment]
image
test at my local environment succeeds but not at circleCi,, need to figure out why

File: storage/database/db_manager.go [hunk 2]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+		logger.Crit("Failed to encode bad blocks", "err", err)
+		return
+	}
+	if err := db.Put(badBlockKey, data); err != nil {
+		logger.Crit("Failed to write bad blocks", "err", err)
+		return
+	}
+}
+
+// DeleteBadBlocks deletes all the bad blocks from the database. Not used anywhere
```
Comment (hyunsooda) [inline storage/database/db_manager.go:1540]: How about relocating this definition to the test file?

File: blockchain/blockchain.go (hunk 1 above)
Comment (hyunsooda) [reply blockchain/blockchain.go:2395]: @ian0371's suggestion seems better in my view. If the string is empty, the identification of empty is more easy thanks to the header (`receipt` in the suggested log)

File: node/sc/bridge_manager.go [hunk 3]
```diff
@@ -194,6 +194,7 @@ func (bi *BridgeInfo) GetCounterPartToken(token common.Address) common.Address {
 
 func (bi *BridgeInfo) GetPendingRequestEvents() []IRequestValueTransferEvent {
 	ready := bi.pendingRequestEvent.Pop(maxPendingNonceDiff / 2)
+
```
Comment (hyunsooda) [inline node/sc/bridge_manager.go:197]: This change seems not related to the PR's changes

File: blockchain/blockchain.go [hunk 4]
```diff
@@ -2362,57 +2357,42 @@ type BadBlockArgs struct {
...
-			block := blk.(*types.Block)
-			blocks = append(blocks, BadBlockArgs{block.Hash(), block})
-		} else {
-			cacheGetBadBlockMissMeter.Mark(1)
-		}
+	blocks, err := bc.db.ReadAllBadBlocks()
+	if err != nil {
+		return nil, err
+	}
+	badblockargs := make([]BadBlockArgs, len(blocks))
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply blockchain/blockchain.go:2364]: I'll apply your suggestion. thanks

File: blockchain/blockchain.go (hunk 1 above)
Comment (hqjang-pepper) [reply blockchain/blockchain.go:2395]: @ian0371 @hyunsooda Got it. Apply your suggestion.

File: node/sc/bridge_manager.go (hunk 3 above)
Comment (hqjang-pepper) [reply node/sc/bridge_manager.go:197]: Thanks. I'll apply.

File: storage/database/db_manager.go [hunk 5]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+type badBlockList []*badBlock
+
+func (s badBlockList) Len() int { return len(s) }
+func (s badBlockList) Less(i, j int) bool {
+	return s[i].Header.Number.Uint64() < s[j].Header.Number.Uint64()
+}
+func (s badBlockList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+// ReadBadBlock retrieves the bad block with the corresponding block hash.
+func (dbm *databaseManager) ReadBadBlock(hash common.Hash) *types.Block {
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1462]: We don't use `Find` for prefix at dbManager. Instead we use `get` or `read` prefix.

File: storage/database/db_manager.go [hunk 6]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+	var badBlocks badBlockList
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		return nil, err
+	}
+	if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+		return nil, err
+	}
+	blocks := make([]*types.Block, 0)
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1492]: I was to, but prealloc problem occured when my code was like your suggestion.

File: storage/database/db_manager.go [hunk 7]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+	for _, bad := range badBlocks {
+		blocks = append(blocks, types.NewBlockWithHeader(bad.Header).WithBody(bad.Body.Transactions))
+	}
+	return blocks, nil
+}
+
+// WriteBadBlock serializes the bad block into the database. If the cumulated
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1499]: You got the wrong point. `WriteBadBlock` method writes single block when it succeeds. Check the code again.

File: storage/database/db_manager.go [hunk 8]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+// bad blocks exceeds the limitation, the oldest will be dropped.
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1500]: exceeds -> exceed change applied.
I think the word `pool` is vague in database. Instead we use capacity or limitation, so I think my comment seems not wrong.

File: storage/database/db_manager.go [hunk 9]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+func (dbm *databaseManager) WriteBadBlock(block *types.Block) {
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1501]: WriteBadBlock writes single block if succeeds.

File: storage/database/db_manager.go [hunk 10]
```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+			logger.Error("failed to decode old bad blocks")
+		}
+	}
+
+	for _, badblock := range badBlocks {
+		if badblock.Header.Hash() == block.Hash() && badblock.Header.Number.Uint64() == block.NumberU64() {
+			logger.Info("There is already corresponding badblock in db.", "badblock number", block.NumberU64())
+			return
+		}
+	}
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1519]: Even if we just return, the badblocks that have already been saved are still stored in the db. And that case is included in the testcode.
```
// block #1 insert again
		dbm.WriteBadBlock(block)
		badBlocks, _ := dbm.ReadAllBadBlocks()
		if len(badBlocks) != 2 {
			t.Fatalf("bad block db len mismatching, have %d, want %d", len(badBlocks), 2)
		}
```

File: storage/database/db_manager.go (hunk 2 above)
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1540]: Umm, why? This comment is about definition of DeleteBadBlocks.

File: blockchain/blockchain.go (hunk 1 above)
Comment (ian0371) [reply blockchain/blockchain.go]: @hqjang-pepper  Actually, I meant logging without `Sprintf()`, but passing variables in the logger's format, i.e., `logger.Error("message", "var1", var1)`.

File: storage/database/db_manager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (hyunsooda) [reply storage/database/db_manager.go:1492]: Okay, then how about this one? `blocks := make([]*types.Block, len(badBlocks))`

File: storage/database/db_manager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hyunsooda) [reply storage/database/db_manager.go:1462]: Okay.

File: storage/database/db_manager.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (hyunsooda) [reply storage/database/db_manager.go:1499]: Is not the `data` a list of badblock? the badblocks are encoded at line 1529 and are inserted to DB at line 1534 .

File: storage/database/db_manager.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (hyunsooda) [reply storage/database/db_manager.go]: Okay. `capacity` sounds good to me

File: storage/database/db_manager.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (hyunsooda) [reply storage/database/db_manager.go:1501]: A list of bad blocks seems to be inserted at line 1529 and 1534.

File: storage/database/db_manager.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hyunsooda) [reply storage/database/db_manager.go:1519]: Now I see. Let me ask another question.
- If a `block` may be clipped by sorting. To guarantee the new bad block must be alive after doing sort, the header number must be comparably higher to be not evicted by sort. There's no related condition inside of function even the near call site. 
- Furthermore, the bad blocks removed by the capacity 100 (`badBlockToKeep`) are now lost. Could you explain why its removal is introduced? (i.e., Your change says that no need to do manage exceeding 100 bad blocks)

File: storage/database/db_manager.go (hunk 2 above)
Comment (hyunsooda) [reply storage/database/db_manager.go:1540]: I meant the definition is a function definition. I think it seems better to relocate this definition to test file since this function is for a test purpose.

File: blockchain/blockchain.go (hunk 1 above)
Comment (hyunsooda) [reply blockchain/blockchain.go]: Please check out the suggested log again. The quality of readability seems not improved with formatted logging by `fmt.Sprintf`

File: storage/database/db_manager.go [hunk 11]
```diff
@@ -1436,6 +1441,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
 	dbm.cm.deleteBlockCache(hash)
 }
 
+const badBlockToKeep = 100
```
Comment (yoomee1313) [inline storage/database/db_manager.go:1444]: Before the change, it has `maxBadBlocks` variable as a limitation, however, you changed to store bad blocks in db. Is this limitation still needed?

File: blockchain/blockchain.go [hunk 12]
```diff
@@ -187,8 +186,6 @@ type BlockChain struct {
 	validator  Validator  // block and state validator interface
 	vmConfig   vm.Config
 
-	badBlocks *lru.Cache // Bad block cache
```
Comment (yoomee1313) [inline blockchain/blockchain.go:190]: Can I ask why you change the bad block storing method from cache to db?

File: storage/database/db_manager.go [hunk 13]
```diff
@@ -1436,6 +1441,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+		logger.Crit("Failed to encode bad blocks", "err", err)
+		return
+	}
+	if err := db.Put(badBlockKey, data); err != nil {
+		logger.Crit("Failed to write bad blocks", "err", err)
+		return
+	}
+}
+
+func (dbm *databaseManager) DeleteBadBlocks() {
```
Comment (yoomee1313) [inline storage/database/db_manager.go:1540]: Is it likely to be used in future work? If so, what about putting TODO?

File: storage/database/db_manager.go (hunk 2 above)
Comment (yoomee1313) [reply storage/database/db_manager.go]: @hqjang-pepper Is it likely to be used in future work? If so, what about putting TODO? If it is only used at test file only, i also think moving to the test file of deleting is better as @hyunsooda said.

File: blockchain/blockchain.go (hunk 12 above)
Comment (hqjang-pepper) [reply blockchain/blockchain.go:190]: Ethereum also stores badblocks into DB now.
Plus if the memory runs out when a node dies, badblocks cannot be retrieved, which makes it impossible to debug badblocks when network goes down.

File: storage/database/db_manager.go (hunk 11 above)
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1444]: Good suggestion. Lake also pointed out this part, and it is under discussion.

File: storage/database/db_manager.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1519]: Good suggestion. under discussion.

File: blockchain/blockchain.go (hunk 12 above)
Comment (yoomee1313) [reply blockchain/blockchain.go:190]: @hqjang-pepper Oh, I see. Is there any related ethereum PR? It would be helpful to understand the history.

Comment (hqjang-pepper) [reply blockchain/blockchain.go:190]: Checkout this PR : https://github.com/ethereum/go-ethereum/pull/21827

File: storage/database/db_manager.go [hunk 14]
```diff
@@ -1448,6 +1453,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+func (dbm *databaseManager) ReadAllBadBlocks() ([]*types.Block, error) {
+	var badBlocks badBlockList
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		return nil, err
+	}
+	if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+		return nil, err
+	}
```
Comment (kjhman21) [inline storage/database/db_manager.go:1503]: Seems like reading bad blocks from DB is all the same. Can we make an internal function for this? If we do that, we can do refactor code more easily if needed.

**Review by aidan-kwon — DISMISSED** [review]

**Comment by aidan-kwon** [comment]
@hqjang-pepper Could you fix test failures?

**Review by aidan-kwon — DISMISSED** [review]

**Review by yoomee1313 — DISMISSED** [review]

**Review by hyunsooda — COMMENTED** [review]
How did you decide on the space constraint of bad blocks? If it can be possibly leveraged for debug purposes, my suggestion is not to make the constraint of storage space for badblocks.

File: storage/database/db_manager.go [hunk 15]
```diff
@@ -1448,6 +1453,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+func (dbm *databaseManager) WriteBadBlock(block *types.Block) {
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		logger.Warn("Failed to load old bad blocks", "error", err)
+	}
+	var badBlocks badBlockList
+	if len(blob) > 0 {
+		if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+			logger.Error("failed to decode old bad blocks")
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1522]: Good suggestion. I'll apply that.

File: storage/database/db_manager.go [hunk 16]
```diff
@@ -1448,6 +1453,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+	blocks := make([]*types.Block, len(badBlocks))
+	for i, bad := range badBlocks {
+		blocks[i] = types.NewBlockWithHeader(bad.Header).WithBody(bad.Body.Transactions)
+	}
+	return blocks, nil
+}
+
+// WriteBadBlock serializes the bad block into the database. If the cumulated
+// bad blocks exceed the capacity, the oldest will be dropped.
+func (dbm *databaseManager) WriteBadBlock(block *types.Block) {
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1513]: The reason to read all badblocks is to store badblocks in order. I agree this operation itself is costly high, but since max length is 100, it's not a big deal.

File: storage/database/db_manager.go [hunk 17]
```diff
@@ -1448,6 +1453,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
...
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		logger.Warn("Failed to load old bad blocks", "error", err)
+	}
+	var badBlocks badBlockList
+	if len(blob) > 0 {
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1520]: If len(blob) ==0, it means that badblocklist is empty, so we don't have to execute `DecodeBytes` logic. I think this is a shortcut for 'no badblock case'. Agree?

File: storage/database/db_manager_test.go [hunk 18]
```diff
@@ -351,6 +352,73 @@ func TestDBManager_Block(t *testing.T) {
...
+		// block #1 insert again
+		dbm.WriteBadBlock(block)
+		badBlocks, _ := dbm.ReadAllBadBlocks()
+		if len(badBlocks) != 2 {
+			t.Fatalf("bad block db len mismatching, have %d, want %d", len(badBlocks), 2)
+		}
+
+		// Write a bunch of bad blocks, all the blocks are should sorted
+		// in reverse order. The extra blocks should be truncated.
+		for _, n := range rand.Perm(150) {
```
Comment (aidan-kwon) [inline storage/database/db_manager_test.go:397]:
```suggestion
		for _, n := range rand.Perm(110) {
```

Comment (aidan-kwon) [reply storage/database/db_manager_test.go]: BadgerDB fails to open a new database in another TC if this test writes some amount of data. I don't know the exact reason for it yet, but I don't this the reason is related to this PR. So, I just reduce the size of stored data in this TC. 
It would be better if you investigate more why BadgerDB failed with large data after merging this PR.

**Review by aidan-kwon — APPROVED** [review]

File: storage/database/db_manager_test.go (hunk 18 above)
Comment (hqjang-pepper) [reply storage/database/db_manager_test.go]: I'll make an issue. Thanks.

**Review by jeongkyun-oh — APPROVED** [review]

File: storage/database/db_manager.go (hunk 14 above)
Comment (hqjang-pepper) [reply storage/database/db_manager.go:1503]: Because the return arguments of readbadbock and readallbadblock are 1 and 2 respectively,
Even if the logic is the same, it is difficult to tie these parts in common.
I think it'll be fine to leave as it is.
